src/posix/platform/multithread_api.cpp

```diff
@@ -0,0 +1,265 @@
+
+    platformConfig.mInterfaceName = iface;
+    VerifyOrDie(platformConfig.mRadioUrlNum < OT_ARRAY_LENGTH(platformConfig.mRadioUrls), OT_EXIT_INVALID_ARGUMENTS);
+    platformConfig.mRadioUrls[platformConfig.mRadioUrlNum++] = radioUrl;
+#if defined(__linux)
+    platformConfig.mRealTimeSignal = 41;
+#endif
+    platformConfig.mSpeedUpFactor = 1;
+
+    syslog(LOG_INFO, "Running %s", otGetVersionString());
+    syslog(LOG_INFO, "Thread version: %hu", otThreadGetVersion());
+    IgnoreError(otLoggingSetLevel(config->mOtLogLevel));
+
+    gInstance = otSysInit(&platformConfig);
+    if (gInstance == nullptr)
+    {
+        syslog(LOG_INFO, "otSysInit failed");
+        DieNow(OT_EXIT_FAILURE);
+    }
+    syslog(LOG_INFO, "Thread interface: %s", otSysGetThreadNetifName());
+    syslog(LOG_INFO, "RCP version: %s", otPlatRadioGetVersionString(gInstance));
+
+    otLogDebgPlat("ot instance create success!!!");
+
+    pthread_cond_signal(&gCond);
+    pthread_mutex_unlock(&gLock);
+
+    while (gTerminate == false)
+    {
+        otSysMainloopContext mainloop;
+
+        otTaskletsProcess(gInstance);
+
+        FD_ZERO(&mainloop.mReadFdSet);
+        FD_ZERO(&mainloop.mWriteFdSet);
+        FD_ZERO(&mainloop.mErrorFdSet);
+
+        mainloop.mMaxFd           = -1;
+        mainloop.mTimeout.tv_sec  = 10;
+        mainloop.mTimeout.tv_usec = 0;
+
+        otSysMainloopUpdate(gInstance, &mainloop);
+
+        if (otSysMainloopPoll(&mainloop) >= 0)
+        {
+            pthread_mutex_lock(&gLock);
+
+            otSysMainloopProcess(gInstance, &mainloop);
+
+            pthread_mutex_unlock(&gLock);
+        }
+        else if (errno != EINTR)
+        {
+            perror("select");
+            ExitNow();
+        }
+    }
+
+exit:
+    if (gInstance)
+    {
+        otSysDeinit();
+    }
+    gThreadId  = 0;
+    gInstance  = nullptr;
+    gTerminate = false;
+    pthread_cond_signal(&gCond);
+    pthread_mutex_unlock(&gLock);
+    syslog(LOG_INFO, "terminate thread mainloop : exit");
+    return nullptr;
+}
+
+void otSysGetInstance(otInstance **aInstance, const char *aComPort, uint16_t aOtLogLevel)
+{
+    OtConfig *config = nullptr;
+
+    syslog(LOG_INFO, "otSysGetInstance");
+
+    if (gInstance)
+    {
+        syslog(LOG_INFO, "ot instance already initialised!!!");
+        *aInstance = gInstance;
+        ExitNow();
+    }
+
+    config = (OtConfig *)malloc(sizeof(OtConfig));
+    VerifyOrExit(config != nullptr, syslog(LOG_INFO, "malloc failed"));
+    memset(config, 0, sizeof(OtConfig));
+    strncpy(config->mComPort, aComPort, strlen(aComPort));
+    config->mOtLogLevel = aOtLogLevel;
+
+    VerifyOrExit(pthread_mutex_init(&gLock, NULL) == 0, syslog(LOG_INFO, "mutex init has failed"));
+    VerifyOrExit(pthread_cond_init(&gCond, NULL) == 0, syslog(LOG_INFO, "cond init has failed"));
+
+    pthread_create(&gThreadId, NULL, pThreadOtMainLoop, config);
+    pthread_mutex_lock(&gLock);
```

Comment:
I think we need to set `config = null` here after passing it `pthread_create ()` so to not `free()` it at exit?

Reply:
In old commit, we passed the config to `pThreadOtMainLoop` and then the memory of config was released in `pThreadOtMainLoop`. 
But, in latest commit, this logic has changed.
`pThreadOtMainLoop` uses the config passed and after that, it sends the signal to the caller of pthread_create. 
There is no `free()` in `pThreadOtMainLoop` now.
After getting this signal, `otSysGetInstance` releases the memory of config at `exit`.

So, I think we should not set `config = null` here. 
How about your opinion ?

Reply:
I see. got it. Sounds good. Thanks.